exec/java-exec/src/main/java/org/apache/drill/exec/util/filereader/BufferedDirectBufInputStream.java

```diff
@@ -0,0 +1,467 @@
+   */
+  protected volatile DrillBuf internalBuffer; // the internal buffer
+
+  /**
+   * The number of valid bytes in <code>internalBuffer</code>.
+   * <code> count </code> is always in the range <code>[0,internalBuffer.capacity]</code>
+   * <code>internalBuffer[count-1]</code> is the last valid byte in the buffer.
+   */
+  protected int count;
+
+  /**
+   * The current read position in the buffer; the index of the next
+   * character to be read from the <code>internalBuffer</code> array.
+   * <p/>
+   * This value is always in the range <code>[0,count]</code>.
+   * If <code>curPosInBuffer</code> is equal to <code>count></code> then we have read
+   * all the buffered data and the next read (or skip) will require more data to be read
+   * from the underlying input stream.
+   */
+  protected int curPosInBuffer;
+
+  protected long curPosInStream; // current offset in the input stream
+
+  private int bufSize;
+
+  private volatile DrillBuf tempBuffer; // a temp Buffer for use by read(byte[] buf, int off, int len)
+
+  private DrillBuf getBuf() throws IOException {
+    checkInputStreamState();
+    if (internalBuffer == null) {
+      throw new IOException("Input stream is closed.");
+    }
+    return this.internalBuffer;
+  }
+
+  /**
+   * Creates a <code>BufferedDirectBufInputStream</code>
+   * with the default (8 MiB) buffer size.
+   */
+  public BufferedDirectBufInputStream(InputStream in, BufferAllocator allocator, String id,
+      long startOffset, long totalByteSize, boolean enableHints) {
+    this(in, allocator, id, startOffset, totalByteSize, defaultBufferSize, enableHints);
+  }
+
+  /**
+   * Creates a <code>BufferedDirectBufInputStream</code>
+   * with the specified buffer size.
+   */
+  public BufferedDirectBufInputStream(InputStream in, BufferAllocator allocator, String id,
+      long startOffset, long totalByteSize, int bufSize, boolean enableHints) {
+    super(in, allocator, id, startOffset, totalByteSize, enableHints);
+    Preconditions.checkArgument(bufSize >= 0);
+    // We make the buffer size the smaller of the buffer Size parameter or the total Byte Size
+    // rounded to next highest pwoer of two
+    int bSize = bufSize < (int) totalByteSize ? bufSize : (int) totalByteSize;
+    // round up to next power of 2
+    bSize--;
+    bSize |= bSize >>> 1;
+    bSize |= bSize >>> 2;
+    bSize |= bSize >>> 4;
+    bSize |= bSize >>> 8;
+    bSize |= bSize >>> 16;
+    bSize++;
+    this.bufSize = bSize;
+
+  }
+
+  @Override public void init() throws UnsupportedOperationException, IOException {
+    super.init();
+    this.internalBuffer = this.allocator.buffer(this.bufSize);
+    this.tempBuffer = this.allocator.buffer(defaultTempBufferSize);
+    int bytesRead = getNextBlock();
+    if (bytesRead <= 0) {
+      throw new IOException("End of stream reached while initializing buffered reader.");
+    }
+  }
+
+  private DrillBuf reallocBuffer(int newSize ){
```

Comment:
realloc seem to imply we copy the old data which is not the case here. Should we name this function better ?